### PR TITLE
ISPN-7714 Extend server configuration to allow easier deployment of user code

### PR DESCRIPTION
**Pull request by anistor**
In this particular case it is just the analysers, but it can be anything else using classpath discovery via java.util.ServiceLoader mechanism or another configuration file.   

https://issues.jboss.org/browse/ISPN-7714

### CODE REVIEW

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerAddHandler.java [hunk 1]
```diff
@@ -171,8 +171,19 @@ static void installRuntimeServices(OperationContext context, ModelNode operation
                     }
                     authorizationConfig.getRoles().put(roleName, permissions);
                 }
+            }
+        }
 
+        if (model.hasDefined(ModelKeys.MODULES) && model.get(ModelKeys.MODULES).hasDefined(ModelKeys.MODULES_NAME)) {
+            ModelNode modulesModel = model.get(ModelKeys.MODULES, ModelKeys.MODULES_NAME);
+            LinkedList<ModuleIdentifier> modules = new LinkedList<>();
```
Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerAddHandler.java]: Any reason this is a `LinkedList`? We could just initialize this as an `ArrayList` with size equal to the size of the list returned from `asList`.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java [hunk 2]
```diff
@@ -116,6 +120,18 @@ public ServiceName getServiceName() {
                 .addDependency(ScheduledThreadPoolResource.EXPIRATION.getServiceName(this.name), ThreadPoolConfiguration.class, this.expirationThreadPool)
                 .addDependency(ScheduledThreadPoolResource.REPLICATION_QUEUE.getServiceName(this.name), ThreadPoolConfiguration.class, this.replicationQueueThreadPool)
         ;
+        if (module != null) {
+            if (!module.getName().equals("org.infinispan.extension")) {
+                // todo [anistor] only works for dynamic modules
```
Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: I noticed these similar todos around. Is this something we need to do in this PR or do we need a different JIRA?

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java [hunk 3]
```diff
@@ -229,11 +245,34 @@ public GlobalConfiguration getValue() {
         return builder.build();
     }
 
+    private ClassLoader makeAggregatedClassLoader(ModuleLoader moduleLoader, ModuleIdentifier cacheContainerModule, List<ModuleIdentifier> additionalModules) throws ModuleLoadException {
+        if (additionalModules != null) {
+            List<ClassLoader> classLoaders = new ArrayList<>();
```
Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: We could make this a `ClassLoader[]` that is initialized to `additionalModules.size` +1 if `cacheContainerModule` wasn't null.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java [hunk 4]
```diff
@@ -245,6 +245,14 @@ private void parseContainer(XMLExtendedStreamReader reader, PathAddress subsyste
                    }
                    break;
                 }
+                case MODULES: {
+                   if (namespace.since(Namespace.INFINISPAN_SERVER_9_1)) {
```
Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java]: Should be 9_2 right?

Comment (anistor) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java]: indeed. solved.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java (hunk 2 above)
Comment (anistor) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: Yes, this should be a different JIRA, not reported yet. Will report it and update the comments with the id.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerAddHandler.java (hunk 1 above)
Comment (anistor) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerAddHandler.java]: no particular reason. suggestion applied.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java (hunk 3 above)
Comment (anistor) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: Both could be null. Creating a list that will contain ```module``` + ```modules``` will not look pretty, so I preferred to pass both ```module``` and ```modules``` as separate args to ```makeAggregatedClassLoader```.

**Comment by anistor** [comment]
updated and rebased.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java (hunk 3 above)
Comment (wburns) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: `additionalModules` in this section was already checked against null. But I am not picky about this, I am fine the way it is.

**Comment by wburns** [comment]
It looks like this is having issues with finishing the build

http://ci.infinispan.org/job/Infinispan/job/PR-5535/8/console

Postponing this until next build.

**Comment by wburns** [comment]
It looks like the failures are no on master as well, so the issue seems unrelated.

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java (hunk 3 above)
Comment (anistor) [reply server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CacheContainerConfigurationBuilder.java]: I do not understand what you mean.

**Comment by anistor** [comment]
rebased again

**Comment by wburns** [comment]
This needs rebase so it can pass.

**Comment by anistor** [comment]
rebased

**Comment by anistor** [comment]
I have updates some test again.

**Comment by gustavonalle** [comment]
Integrated, thanks @anistor

**Pull request merged**